SubprocessReportGenerator failure to refresh report for active recording

Since #359 

```
INFO: (127.0.0.1:37548): GET /api/v1/targets/service:jmx:rmi:%2F%2F%2Fjndi%2Frmi:%2F%2Fcontainer-jfr:9093%2Fjmxrmi/reports/foo 200 6ms
Dec 21, 2020 4:51:22 PM com.redhat.rhjmc.containerjfr.core.log.Logger info
INFO: com.redhat.rhjmc.containerjfr.net.reports.SubprocessReportGenerator starting
Dec 21, 2020 4:51:22 PM com.redhat.rhjmc.containerjfr.core.log.Logger info
INFO: com.redhat.rhjmc.containerjfr.net.reports.SubprocessReportGenerator adjusting subprocess OOM score
Dec 21, 2020 4:51:22 PM com.redhat.rhjmc.containerjfr.core.log.Logger info
INFO: com.redhat.rhjmc.containerjfr.net.reports.SubprocessReportGenerator processing report
Dec 21, 2020 4:51:23 PM com.redhat.rhjmc.containerjfr.core.log.Logger info
INFO: com.redhat.rhjmc.containerjfr.net.reports.SubprocessReportGenerator writing report to file
Dec 21, 2020 4:51:23 PM com.github.benmanes.caffeine.cache.BoundedLocalCache lambda$refreshIfNeeded$5
WARNING: Exception thrown during refresh
java.util.concurrent.CompletionException: java.nio.file.NoSuchFileException: /tmp/3117503518209689085.tmp
	at com.github.benmanes.caffeine.cache.CacheLoader.lambda$asyncReload$2(CacheLoader.java:194)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1692)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.nio.file.NoSuchFileException: /tmp/3117503518209689085.tmp
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:370)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:421)
	at java.base/java.nio.file.Files.readAllBytes(Files.java:3205)
	at java.base/java.nio.file.Files.readString(Files.java:3283)
	at java.base/java.nio.file.Files.readString(Files.java:3242)
	at com.redhat.rhjmc.containerjfr.core.sys.FileSystem.readString(FileSystem.java:91)
	at com.redhat.rhjmc.containerjfr.net.reports.ActiveRecordingReportCache.getReport(ActiveRecordingReportCache.java:140)
	at com.redhat.rhjmc.containerjfr.net.reports.ActiveRecordingReportCache.lambda$new$0(ActiveRecordingReportCache.java:102)
	at com.github.benmanes.caffeine.cache.CacheLoader.reload(CacheLoader.java:166)
	at com.github.benmanes.caffeine.cache.CacheLoader.lambda$asyncReload$2(CacheLoader.java:190)
	... 7 more
```

--- a/src/main/java/com/redhat/rhjmc/containerjfr/net/reports/ActiveRecordingReportCache.java
+++ b/src/main/java/com/redhat/rhjmc/containerjfr/net/reports/ActiveRecordingReportCache.java
@@ -43,7 +43,6 @@ package com.redhat.rhjmc.containerjfr.net.reports;
 
 import java.nio.file.Path;
 import java.time.Duration;
-import java.util.Objects;
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionException;
@@ -54,9 +53,6 @@ import java.util.concurrent.locks.ReentrantLock;
 
 import javax.inject.Named;
 import javax.inject.Provider;
-
-import org.apache.commons.lang3.builder.EqualsBuilder;
-import org.apache.commons.lang3.builder.HashCodeBuilder;
 
 import org.openjdk.jmc.rjmx.services.jfr.IRecordingDescriptor;
 
@@ -69,6 +65,7 @@ import com.redhat.rhjmc.containerjfr.core.sys.FileSystem;
 import com.redhat.rhjmc.containerjfr.net.ConnectionDescriptor;
 import com.redhat.rhjmc.containerjfr.net.TargetConnectionManager;
 import com.redhat.rhjmc.containerjfr.net.reports.SubprocessReportGenerator.ExitStatus;
+import com.redhat.rhjmc.containerjfr.net.reports.SubprocessReportGenerator.RecordingDescriptor;
 import com.redhat.rhjmc.containerjfr.net.reports.SubprocessReportGenerator.ReportGenerationException;
 import com.redhat.rhjmc.containerjfr.net.web.http.generic.TimeoutHandler;
 
@@ -174,42 +171,6 @@ class ActiveRecordingReportCache {
             if (saveFile != null) {
                 fs.deleteIfExists(saveFile);
             }
-        }
-    }
-
-    static class RecordingDescriptor {
-        final ConnectionDescriptor connectionDescriptor;
-        final String recordingName;
-
-        RecordingDescriptor(ConnectionDescriptor connectionDescriptor, String recordingName) {
-            this.connectionDescriptor = Objects.requireNonNull(connectionDescriptor);
-            this.recordingName = Objects.requireNonNull(recordingName);
-        }
-
-        @Override
-        public boolean equals(Object other) {
-            if (other == null) {
-                return false;
-            }
-            if (other == this) {
-                return true;
-            }
-            if (!(other instanceof RecordingDescriptor)) {
-                return false;
-            }
-            RecordingDescriptor rd = (RecordingDescriptor) other;
-            return new EqualsBuilder()
-                    .append(connectionDescriptor, rd.connectionDescriptor)
-                    .append(recordingName, rd.recordingName)
-                    .isEquals();
-        }
-
-        @Override
-        public int hashCode() {
-            return new HashCodeBuilder()
-                    .append(connectionDescriptor)
-                    .append(recordingName)
-                    .hashCode();
         }
     }
 }

--- a/src/main/java/com/redhat/rhjmc/containerjfr/net/reports/SubprocessReportGenerator.java
+++ b/src/main/java/com/redhat/rhjmc/containerjfr/net/reports/SubprocessReportGenerator.java
@@ -63,6 +63,9 @@ import java.util.concurrent.TimeUnit;
 
 import javax.inject.Provider;
 
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
 import org.openjdk.jmc.rjmx.ConnectionException;
 import org.openjdk.jmc.rjmx.services.jfr.IRecordingDescriptor;
 
@@ -75,7 +78,6 @@ import com.redhat.rhjmc.containerjfr.core.sys.Environment;
 import com.redhat.rhjmc.containerjfr.core.sys.FileSystem;
 import com.redhat.rhjmc.containerjfr.net.ConnectionDescriptor;
 import com.redhat.rhjmc.containerjfr.net.TargetConnectionManager;
-import com.redhat.rhjmc.containerjfr.net.reports.ActiveRecordingReportCache.RecordingDescriptor;
 import com.redhat.rhjmc.containerjfr.net.reports.ReportService.RecordingNotFoundException;
 import com.redhat.rhjmc.containerjfr.util.JavaProcess;
 
@@ -337,6 +339,42 @@ public class SubprocessReportGenerator {
         } catch (IOException ioe) {
             ioe.printStackTrace();
             throw new ReportGenerationException(ExitStatus.IO_EXCEPTION);
+        }
+    }
+
+    static class RecordingDescriptor {
+        final ConnectionDescriptor connectionDescriptor;
+        final String recordingName;
+
+        RecordingDescriptor(ConnectionDescriptor connectionDescriptor, String recordingName) {
+            this.connectionDescriptor = Objects.requireNonNull(connectionDescriptor);
+            this.recordingName = Objects.requireNonNull(recordingName);
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (other == null) {
+                return false;
+            }
+            if (other == this) {
+                return true;
+            }
+            if (!(other instanceof RecordingDescriptor)) {
+                return false;
+            }
+            RecordingDescriptor rd = (RecordingDescriptor) other;
+            return new EqualsBuilder()
+                    .append(connectionDescriptor, rd.connectionDescriptor)
+                    .append(recordingName, rd.recordingName)
+                    .isEquals();
+        }
+
+        @Override
+        public int hashCode() {
+            return new HashCodeBuilder()
+                    .append(connectionDescriptor)
+                    .append(recordingName)
+                    .hashCode();
         }
     }
 

--- a/src/main/java/com/redhat/rhjmc/containerjfr/net/reports/SubprocessReportGenerator.java
+++ b/src/main/java/com/redhat/rhjmc/containerjfr/net/reports/SubprocessReportGenerator.java
@@ -171,16 +171,16 @@ public class SubprocessReportGenerator {
     }
 
     Future<Path> exec(RecordingDescriptor recordingDescriptor, Duration timeout) throws Exception {
-        // TODO add a FileSystem abstraction around Files.createTemp*
         Path recording =
                 getRecordingFromLiveTarget(
                         recordingDescriptor.recordingName,
                         recordingDescriptor.connectionDescriptor);
-        CompletableFuture<Path> cf = exec(recording, tempFileProvider.get(), timeout);
-        cf.whenCompleteAsync(
+        Path saveFile = tempFileProvider.get();
+        CompletableFuture<Path> cf = exec(recording, saveFile, timeout);
+        cf.whenComplete(
                 (p, t) -> {
                     try {
-                        fs.deleteIfExists(p);
+                        fs.deleteIfExists(recording);
                     } catch (IOException e) {
                         logger.warn(e);
                     }

--- a/src/test/java/com/redhat/rhjmc/containerjfr/net/reports/ActiveRecordingReportCacheTest.java
+++ b/src/test/java/com/redhat/rhjmc/containerjfr/net/reports/ActiveRecordingReportCacheTest.java
@@ -68,9 +68,9 @@ import com.redhat.rhjmc.containerjfr.core.sys.Environment;
 import com.redhat.rhjmc.containerjfr.core.sys.FileSystem;
 import com.redhat.rhjmc.containerjfr.net.ConnectionDescriptor;
 import com.redhat.rhjmc.containerjfr.net.TargetConnectionManager;
-import com.redhat.rhjmc.containerjfr.net.reports.ActiveRecordingReportCache.RecordingDescriptor;
 import com.redhat.rhjmc.containerjfr.net.reports.ReportService.RecordingNotFoundException;
 import com.redhat.rhjmc.containerjfr.net.reports.SubprocessReportGenerator.ExitStatus;
+import com.redhat.rhjmc.containerjfr.net.reports.SubprocessReportGenerator.RecordingDescriptor;
 import com.redhat.rhjmc.containerjfr.util.JavaProcess;
 
 @ExtendWith(MockitoExtension.class)

--- a/src/test/java/com/redhat/rhjmc/containerjfr/net/reports/SubprocessReportGeneratorTest.java
+++ b/src/test/java/com/redhat/rhjmc/containerjfr/net/reports/SubprocessReportGeneratorTest.java
@@ -61,7 +61,9 @@ import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.ArgumentCaptor;
 import org.mockito.Mock;
 import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
 import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.stubbing.Answer;
 
 import com.redhat.rhjmc.containerjfr.core.log.Logger;
 import com.redhat.rhjmc.containerjfr.core.net.Credentials;
@@ -70,8 +72,8 @@ import com.redhat.rhjmc.containerjfr.core.sys.Environment;
 import com.redhat.rhjmc.containerjfr.core.sys.FileSystem;
 import com.redhat.rhjmc.containerjfr.net.ConnectionDescriptor;
 import com.redhat.rhjmc.containerjfr.net.TargetConnectionManager;
-import com.redhat.rhjmc.containerjfr.net.reports.ActiveRecordingReportCache.RecordingDescriptor;
 import com.redhat.rhjmc.containerjfr.net.reports.SubprocessReportGenerator.ExitStatus;
+import com.redhat.rhjmc.containerjfr.net.reports.SubprocessReportGenerator.RecordingDescriptor;
 import com.redhat.rhjmc.containerjfr.util.JavaProcess;
 
 @ExtendWith(MockitoExtension.class)
@@ -86,8 +88,9 @@ class SubprocessReportGeneratorTest {
     ConnectionDescriptor connectionDescriptor;
     RecordingDescriptor recordingDescriptor;
     @Mock Path recordingFile;
-    @Mock Path tempFile;
-    Provider<Path> tempFileProvider;
+    @Mock Path tempFile1;
+    @Mock Path tempFile2;
+    @Mock Provider<Path> tempFileProvider;
     SubprocessReportGenerator generator;
 
     @BeforeEach
@@ -96,9 +99,12 @@ class SubprocessReportGeneratorTest {
                 new ConnectionDescriptor("fooHost:1234", new Credentials("someUser", "somePass"));
         recordingDescriptor = new RecordingDescriptor(connectionDescriptor, "testRecording");
 
-        tempFileProvider = () -> tempFile;
-        Mockito.lenient().when(tempFile.toAbsolutePath()).thenReturn(tempFile);
-        Mockito.lenient().when(tempFile.toString()).thenReturn("/dest/serializedtransformers.tmp");
+        tempFileProvider = Mockito.mock(Provider.class);
+        Mockito.lenient().when(tempFileProvider.get()).thenReturn(tempFile1).thenReturn(tempFile2);
+        Mockito.lenient().when(tempFile1.toAbsolutePath()).thenReturn(tempFile1);
+        Mockito.lenient().when(tempFile1.toString()).thenReturn("/tmp/file1.tmp");
+        Mockito.lenient().when(tempFile2.toAbsolutePath()).thenReturn(tempFile2);
+        Mockito.lenient().when(tempFile2.toString()).thenReturn("/tmp/file2.tmp");
         Mockito.lenient().when(recordingFile.toAbsolutePath()).thenReturn(recordingFile);
         Mockito.lenient().when(recordingFile.toString()).thenReturn("/dest/recording.tmp");
 
@@ -253,6 +259,49 @@ class SubprocessReportGeneratorTest {
                             Matchers.containsString(
                                     "Recording /dest/recording.tmp not found in target archives"));
                 });
+    }
+
+    @Test
+    void shouldExecuteProcessAndDeleteRecordingOnCompletion() throws Exception {
+        Mockito.when(proc.waitFor(10_000, TimeUnit.MILLISECONDS)).thenReturn(true);
+        Mockito.when(proc.exitValue()).thenReturn(ExitStatus.OK.code);
+
+        Mockito.when(targetConnectionManager.executeConnectedTask(Mockito.any(), Mockito.any()))
+                .then(
+                        new Answer<Path>() {
+                            @Override
+                            public Path answer(InvocationOnMock invocation) throws Throwable {
+                                return tempFileProvider.get();
+                            }
+                        });
+
+        Path result = generator.exec(recordingDescriptor, Duration.ofSeconds(10)).get();
+
+        MatcherAssert.assertThat(result, Matchers.sameInstance(tempFile2));
+        Mockito.verify(fs).deleteIfExists(tempFile1);
+    }
+
+    @Test
+    void shouldExecuteProcessAndDeleteRecordingOnFailure() throws Exception {
+        Mockito.when(proc.waitFor(10_000, TimeUnit.MILLISECONDS)).thenReturn(true);
+        Mockito.when(proc.exitValue()).thenReturn(ExitStatus.NO_SUCH_RECORDING.code);
+
+        Mockito.when(targetConnectionManager.executeConnectedTask(Mockito.any(), Mockito.any()))
+                .then(
+                        new Answer<Path>() {
+                            @Override
+                            public Path answer(InvocationOnMock invocation) throws Throwable {
+                                return tempFileProvider.get();
+                            }
+                        });
+
+        Assertions.assertThrows(
+                ExecutionException.class,
+                () -> {
+                    generator.exec(recordingDescriptor, Duration.ofSeconds(10)).get();
+                });
+
+        Mockito.verify(fs).deleteIfExists(tempFile1);
     }
 
     static class TestReportTransformer implements ReportTransformer {

